File: 111.minimum-depth-of-binary-tree.js
/*
 * @lc app=leetcode id=111 lang=javascript
 *
 * [111] Minimum Depth of Binary Tree
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (!root) {
    return 0;
  }
  if (root.left && root.right) {
    return 1 + Math.min(minDepth(root.left), minDepth(root.right));
  }
  return 1 + Math.max(minDepth(root.left), minDepth(root.right));
};
// @lc code=end

File: 111.minimum-depth-of-binary-tree.go
/*
 * @lc app=leetcode id=111 lang=golang
 *
 * [111] Minimum Depth of Binary Tree
 */

package leetcode

// @lc code=start
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	if root.Left == nil {
		return 1 + minDepth(root.Right)
	}
	if root.Right == nil {
		return 1 + minDepth(root.Left)
	}
	return min111(minDepth(root.Left), minDepth(root.Right)) + 1
}

func min111(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @lc code=end
